src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Message\LockableMessageInterface;
+use Symfony\Component\Messenger\Message\TTLAwareLockableMessageInterface;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private ?LockFactory $lockFactory = null;
+
+    public function __construct(
+        ?LockFactory $lockFactory,
+    ) {
+        $this->lockFactory = $lockFactory;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null === $this->lockFactory) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        $message = $envelope->getMessage();
+
+        if (null === $envelope->last(ReceivedStamp::class)) {
+            if ($message instanceof LockableMessageInterface) {
+                // If we're trying to dispatch a lockable message.
+                $keyResource = $message->getKey();
+
+                if (null !== $keyResource) {
+                    $key = new Key($keyResource);
+
+                    // The acquire call must be done before stamping the message
+                    // in order to have the full state of the key in the stamp.
+                    $lock = $message instanceof TTLAwareLockableMessageInterface
+                        ? $this->lockFactory->createLockFromKey($key, $message->getTTL(), autoRelease: false)
+                        : $this->lockFactory->createLockFromKey($key, autoRelease: false);
+
+                    if (!$lock->acquire()) {
+                        return $envelope;
+                    }
+
+                    // The acquire call must be done before stamping the message
+                    // in order to have the full state of the key in the stamp.
+                    $envelope = $envelope->with(new LockStamp($key, $message->shouldBeReleasedBeforeHandlerCall()));
+                }
+            }
+        } else {
+            $this->releaseLock($envelope, true);
+        }
+
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
+            // If we've received a lockable message, we're releasing it.
+            if (null !== $envelope->last(ReceivedStamp::class)) {
```

Comment:
I think this is not necessary, the check is performed in the releaseLock method.
Also, it's not consistent with the call `$this->releaseLock($envelope, true);` at line 64

Reply:
For `$this->releaseLock($envelope, true);` the check on the `ReceivedStamp` is done because I do:
```
if (null === $envelope->last(ReceivedStamp::class)) {
    // foo
} else {
    $this->releaseLock($envelope, true);
}
```

And `releaseLock` is not doing the any check on `ReceivedStamp`, but on `LockStamp`.
If I don't do the `ReceivedStamp` check, I will release the lock right after adding the lock (during the first dispatch) instead of when I receive the message.